plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,300 @@
+        for (MetricType metricType : measurements.keySet()) {
+            if (!topQueriesStores.containsKey(metricType)) {
+                continue;
+            }
+            // add the record to corresponding priority queues to calculate top n queries insights
+            PriorityBlockingQueue<SearchQueryRecord> store = topQueriesStores.get(metricType);
+            this.threadPool.schedule(() -> {
+                checkAndResetWindow(metricType, record.getTimestamp());
+                if (record.getTimestamp() > topQueriesWindowStart.get(metricType)) {
+                    store.add(record);
+                    // remove top elements for fix sizing priority queue
+                    if (store.size() > this.getTopNSize()) {
+                        store.poll();
+                    }
+                }
+            }, delay, QueryInsightsSettings.QUERY_INSIGHTS_EXECUTOR);
+        }
+    }
+
+    private synchronized void checkAndResetWindow(MetricType metricType, Long timestamp) {
+        Long windowStart = calculateWindowStart(timestamp);
+        // reset window if the current window is outdated
+        if (topQueriesWindowStart.get(metricType) < windowStart) {
+            // rotate the current window to history store only if the data belongs to the last window
+            if (topQueriesWindowStart.get(metricType) == windowStart - windowSize.getMillis()) {
+                topQueriesHistoryStores.put(metricType, new ArrayList<>(topQueriesStores.get(metricType)));
+            } else {
+                topQueriesHistoryStores.get(metricType).clear();
+            }
+            topQueriesStores.get(metricType).clear();
+            topQueriesWindowStart.put(metricType, windowStart);
+        }
+    }
+
+    /**
+     * Get all top queries records that are in the current query insight store, based on the input MetricType
+     * Optionally include top N records from the last window.
+     *
+     * By default, return the records in sorted order.
+     *
+     * @param metricType {@link MetricType}
+     * @param includeLastWindow if the top N queries from the last window should be included
+     * @return List of the records that are in the query insight store
+     * @throws IllegalArgumentException if query insight is disabled in the cluster
+     */
+    public List<SearchQueryRecord> getTopNRecords(MetricType metricType, boolean includeLastWindow) throws IllegalArgumentException {
+        if (!enableCollect.containsKey(metricType) || !enableCollect.get(metricType)) {
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ROOT,
+                    "Cannot get query data when query insight feature is not enabled for MetricType [%s].",
+                    metricType
+                )
+            );
+        }
+        checkAndResetWindow(metricType, System.currentTimeMillis());
+        List<SearchQueryRecord> queries = new ArrayList<>(topQueriesStores.get(metricType));
```

Comment:
I'm not sure if it is necessary since we can add attributes in the `SearchQueryRecord` itself. The `topQueriesStores` should be dedicately storing the "top N queries insights". Any other insights should implement their own stores if needed.